### PR TITLE
Ensure Microsoft.NETCore.App is an implicit dependency for publish.

### PR DESCRIPTION
**Pull request by peterhuene**
This commit ensures that we exclude files from the `Microsoft.NETCore.App`
platform library on publish when the platform library is
`Microsoft.AspNetCore.App`, which doesn't have an explicit dependency on
`Microsoft.NETCore.App`.

When publishing a 2.x ASP.NET application, the `MicrosoftNETPlatformLibrary`
property gets changed by ASP.NET to `Microsoft.AspNetCore.App`.  This causes
the task responsible for determining what files to copy locally to treat files
from `Microsoft.NETCore.App` and its dependencies as not being part of the
platform.

However, when publishing the application as framework-dependent, almost all the
platform files get excluded thanks to conflict resolution which has the files
as part of the platform manifest and preferred packages.  Unfortunately, the
apphost, hostpolicy, and hostfxr files are not part of these lists and as a
result will be copied locally.

This breaks framework-dependent apphost activation because hostfxr and
hostpolicy are copied locally.  Additionally, these files end up in the
deps.json file which also prevents roll-forward activation if the two files are
manually deleted.

The fix is to treat `Microsoft.NETCore.App` as an implicit dependency of
`Microsoft.AspNetCore.App` where we calculate the list of excluded packages for
a framework-dependent publish.

Fixes dotnet/cli#10602.

### CODE REVIEW

**Comment by nguerrera** [comment]
I did not review in detail, but can we not consider that there is an implicit dependency from Microsoft.AspNetCore.App to Microsoft.NETCore.App?

I don't understand why this requires hard-coding of file versions, etc.

**Comment by peterhuene** [comment]
The hard coding is effectively what should be present in the platform manifest.

I'm not happy with this solution either.  Let me investigate what it would look like to hardcode an implicit dependency to `Microsoft.NETCore.App` when resolving copy local files for publish.

**Comment by peterhuene** [comment]
Actually, with all the recent changes to asset resolution in 3.0 (namely how we resolve the apphost), this might not need fixing any more in 3.0.  If that's the case, an implicit reference would be ideal vs. needlessly trying to replicate what I envisioned the 3.0 fix would be.

**Comment by peterhuene** [comment]
Ok, I think making an implicit dependency on `Microsoft.NETCore.App` for exclusions is the approach we'll take for this fix.  I'll redo the PR shortly.

**Review by nguerrera — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/ProjectContext.cs
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/ProjectContext.cs]: I think this can also be Microsoft.AspNetCore.*All* on 2.x.

Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ProjectContext.cs]: Maybe consider any platform library that is not base Microsoft.NETCore.App to depend on Microsoft.NETCore.App?

Comment (peterhuene) [reply src/Tasks/Microsoft.NET.Build.Tasks/ProjectContext.cs]: [You are correct](https://github.com/aspnet/Universe/blob/da48de7d7580ce316784dd3e250786ffa7017c25/src/Packages/Microsoft.AspNetCore.All/build/netcoreapp3.0/Microsoft.AspNetCore.All.targets#L10).

Perhaps a check for "not empty and not `Microsoft.NETCore.App`" would be sufficient?

Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ProjectContext.cs]: Yes, that sounds good.

Comment (peterhuene) [reply src/Tasks/Microsoft.NET.Build.Tasks/ProjectContext.cs]: Pushed.

**Review by nguerrera — APPROVED** [review]

**Comment by livarcocc** [comment]
@peterhuene let me know when this is ready for shiproom.

**Comment by peterhuene** [comment]
@livarcocc I think it's ready.  I'll try to be available for the next shiproom to help defend, but I may have to do so from my TEALS class.

**Comment by livarcocc** [comment]
@peterhuene don't worry. I will talk to you tomorrow morning and get the information I need. Don't worry about it when you are in class.

**Comment by peterhuene** [comment]
This has been approved but is waiting for the branch to open for the appropriate release before merging.

**Comment by peterhuene** [comment]
As the 2.2.1xx branch is now open for the next release, I'm merging this.

**Pull request merged**
